File: 73-set-matrix-zeroes/set-matrix-zeroes.cpp
class Solution {
    private:
    void makeZeroes(int row, int col, vector<vector<int>> &matrix, vector<vector<int>> &visited){
        visited[row][col] = 1;
        int n = matrix.size();
        int m = matrix[0].size();

        for (int i = 0;i<m;i++){
            if (matrix[row][i]!=0){
            visited[row][i] = 1;
            matrix[row][i] = 0;
            }
        }
        for (int i = 0;i<n;i++){
            if (matrix[i][col]!=0){
            visited[i][col] = 1;
            matrix[i][col] = 0;
            }
        }
        
    }
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();

        vector<vector<int>> visited(n, vector<int> (m,0));

        for (int i = 0;i<n;i++){
            for (int j = 0;j<m;j++){
                if (matrix[i][j]==0){
                    if (visited[i][j] == 0)
                    makeZeroes(i,j,matrix, visited);


                }
            }
        }

        
    }
};

File: 73-set-matrix-zeroes/set-matrix-zeroes.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m = len(matrix)
        n = len(matrix[0])
        first_col_contain_zero = any(matrix[i][0]==0 for i in range(m)) 
        first_row_contain_zero = any(matrix[0][j]==0 for j in range(n))

        for i in range(1,m):
            for j in range(1,n):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0 
        
        for i in range(1,m):
            for j in range(1,n):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0 

        if first_row_contain_zero:
            for j in range(n):
                matrix[0][j] = 0 
        if first_col_contain_zero:
            for i in range(m):
                matrix[i][0] = 0 
        